library/math/dual.lisp

```diff
@@ -0,0 +1,135 @@
+;;;; dual.lisp
+;;;;
+;;;; dual numbers to enable automatic differentiation
+
+(coalton-library/utils:defstdlib-package #:coalton-library/math/dual
+    (:use
+     #:coalton
+     #:coalton-library/builtin
+     #:coalton-library/classes
+     #:coalton-library/functions
+     #:coalton-library/math/arith
+     #:coalton-library/math/elementary
+     #:coalton-library/math/integral)
+  (:export
+   #:Dual
+   #:primal-part
+   #:dual-part))
+
+(in-package #:coalton-library/math/dual)
+(named-readtables:in-readtable coalton:coalton)
+
+(coalton-toplevel
+
+  (define-type (Dual :t)
+    "Representation of a Dual number in the form (a + bE)."
+    (Dual :t :t))
+  
+  (declare primal-part (Dual :t -> :t))
+  (define (primal-part (Dual p _))
+      p)
+
+  (declare dual-part (Dual :t -> :t))
+  (define (dual-part (Dual _ d))
+      d)
+  
+  (define (sq x)
+      (* x x))
+  
+  (define-instance (Eq :t => Eq (Dual :t))
+    
+    (define (== (Dual a b) (Dual p q))
+      (and (== a p)
+           (== b q))))
+
+  (define-instance (Num :t => Num (Dual :t))
+    
+    (define (+ (Dual p1 d1) (Dual p2 d2))
+      (Dual (+ p1 p2)
+            (+ d1 d2)))
+    
+    (define (- (Dual p1 d1) (Dual p2 d2))
+      (Dual (- p1 p2)
+            (- d1 d2)))
+
+    (define (* (Dual p1 d1) (Dual p2 d2))
+      (Dual (* p1 p2)
+            (+ (* p1  d2) (* d1 p2))))
+    
+    (define (fromInt z)
+      (Dual (fromInt z) (fromInt 0))))
```

Comment:
You can just write `0` here. The `fromInt` call is implicit on integer literals.